### PR TITLE
[infra] Enable libc++ debug mode

### PR DESCRIPTION
**Pull request by jonathanmetzman**
Fixes https://github.com/google/oss-fuzz/issues/610 Fixes https://github.com/google/oss-fuzz/issues/10125

### CODE REVIEW

**Comment by Maximus-** [comment]
You probably want to use -D_LIBCPP_ENABLE_ASSERTIONS (Libc++ "Safe Mode") instead of Libc++ "Debug Mode". 

The debug mode contains checks for iterator invalidation, but from what I can tell, is ABI breaking (even though it tried not to be). I couldn't get it working on some of our code, and I think the Chrome team has had similar experiences. 

The safe mode contains bounds checks and some other sanity, but almost all checks are spatial-safety related.

**Comment by Alan32Liu** [comment]
Trial build failed due to [RPC failure](https://pantheon.corp.google.com/cloud-build/builds;region=us-central1/266f015c-515b-469b-b0fa-1483ab490c9c;step=17?project=oss-fuzz) in `skia`.
Re-run it in case it is flaky.

**Review by Alan32Liu — APPROVED** [review]

**Comment by Alan32Liu** [comment]
> Trial build failed due to [RPC failure](https://pantheon.corp.google.com/cloud-build/builds;region=us-central1/266f015c-515b-469b-b0fa-1483ab490c9c;step=17?project=oss-fuzz) in `skia`. Re-run it in case it is flaky.

Test passed : )

**Comment by Alan32Liu** [comment]
Would there be any comments/suggestions before I merge this? @fanquake

**Comment by jonathanmetzman** [comment]
Don't merge this please. It needs extensive testing, and I need to ask around about it first.

File: infra/base-images/base-clang/checkout_build_install_llvm.sh
Comment (fanquake) [inline infra/base-images/base-clang/checkout_build_install_llvm.sh:101]: Note that if Clang gets rolled to 18.x (#11714) , then we'll want to switch these for `LIBCXX_HARDENING_MODE`: https://libcxx.llvm.org/Hardening.html.

**Pull request closed without merging**
